content/blog/2022-05-19-hcl-mlir/index.md

```diff
@@ -0,0 +1,514 @@
+
+<center>
+<img src="roofline_intro.png" alt="alt_text" title="image_tooltip" style="zoom:20%;">
+</center>
+<center>
+Fig. 5: Roofline Model
+</center>
+
+<!-- What we do to access the memory access / ops -->
+We add a profiling tool to evaluate the operational intensity by analyzing the MLIR code to count the number of memory access and arithmetic operations. Then a roofline model is generated to guide the optimization process. Specifically, we visit each operation and record the current loop nest. When an arithmetic operation or load/store operation is visited, the number of trip counts in the current loop is added. 
+
+Note that we only care about off-chip memory access, so here we need to match the operands of load/store operations with the input arguments and the return argument. This is implemented by comparing the `memref` value. The detailed experiment results are shown in the next section. 
+
+## Experiments
+<!-- A major part of your project is an empirical evaluation. To design your evaluation strategy, you will need to consider at least these things:
+* Where will you get the input code you’ll use in your evaluation?
+* How will you check the correctness of your implementation? If you’ve implemented an optimization, for example, “correctness” means that the transformed programs behave the same way as the original programs.
+* How will you measure the benefit (in performance, energy, complexity, etc.) of your implementation?
+* How will you present the data you collect from your empirical evaluation?
+Other questions may be relevant depending on the project you choose. Consider the [SIGPLAN empirical evaluation guidelines](https://www.sigplan.org/Resources/EmpiricalEvaluation/) when you design your methodology. -->
+
+In this section, we will first talk about the MLIR-based HeteroCL compilation flow and the experimental setup. We will later discuss the results of our memory optimizations.
+
+### MLIR-Based HeteroCL Compilation Flow
+MLIR-based HeteroCL supports two backends for now: a CPU backend through LLVM dialect, and an FPGA backend through Vivado HLS. A HeteroCL program first generates HCL dialect IR, with HCL operations to represent hardware customizations. Then, a transformation pass implements hardware customizations and erases HCL operations, and the IR is converted to affine dialect, as shown in Fig. 6.
+
+<center>
+<img src="hcl-flow.png" alt="alt_text" title="image_tooltip" style="zoom:20%;">
+</center>
+<center>
+Fig. 6: MLIR-based HeteroCL end-to-end compilation flow
+</center>
```

Comment:
This figure refers to something called "SCoP" and "PoCC", which are not defined anywhere, AFAICT?

Reply:
Hi Adrian, you are right. The PoCC branch is for a different project going on in HeteroCL, which is to verify the program is still correct after adding customizations. Since this is not relevant to the memory optimization, I have removed that branch and updated the figure.